typescript/cli/src/validator/preFlightCheck.ts

```diff
@@ -0,0 +1,136 @@
+import { MerkleTreeHook__factory } from '@hyperlane-xyz/core';
+import { HyperlaneCore, S3Validator } from '@hyperlane-xyz/sdk';
+import { Address, eqAddress } from '@hyperlane-xyz/utils';
+
+import { CommandContext } from '../context/types.js';
+import { errorRed, logBlue, logGreen, warnYellow } from '../logger.js';
+
+export const checkValidatorSetup = async (
+  context: CommandContext,
+  chain: string,
+  validators: Set<Address>,
+) => {
+  const { multiProvider, registry } = context;
+
+  const addresses = await registry.getAddresses();
+
+  const core = HyperlaneCore.fromAddressesMap(addresses, multiProvider);
+
+  const validatorAnnounce = core.getContracts(chain).validatorAnnounce;
+  const merkleTreeHook = MerkleTreeHook__factory.connect(
+    addresses[chain].merkleTreeHook,
+    multiProvider.getProvider(chain),
+  );
+
+  let merkleTreeLatestCheckpointIndex: number | undefined;
+  try {
+    const [_, latestCheckpointIndex] = await merkleTreeHook[
+      'latestCheckpoint()'
+    ]();
+
+    merkleTreeLatestCheckpointIndex = latestCheckpointIndex;
+    logBlue(
+      `\nLatest check point index of incremental merkle tree: ${merkleTreeLatestCheckpointIndex}\n`,
+    );
+  } catch (err) {
+    warnYellow(
+      `❗️ Failed to fetch latest checkpoint index of merkleTreeHook on ${chain} \n`,
+    );
+  }
+
+  let announcedValidators;
+  try {
+    announcedValidators = await validatorAnnounce.getAnnouncedValidators();
+  } catch (err) {
+    errorRed(
+      `❌ Failed to fetch announced validators for ${chain}. Exiting.\n`,
+    );
+    process.exit(1);
+  }
+
+  const confirmedValidators = new Set<Address>();
+  const unconfirmedValidators = new Set<Address>();
+
+  for (const validator of validators) {
+    const matches = announcedValidators.filter((address) =>
+      eqAddress(address, validator),
+    );
+    if (matches.length === 0) {
+      unconfirmedValidators.add(validator);
+    } else {
+      confirmedValidators.add(validator);
+    }
+  }
+
+  const errorSet = new Set<string>();
+  errorSet.add('Validator pre flight check failed:');
+
+  if (unconfirmedValidators.size > 0) {
+    errorSet.add('Some validators have not been announced.');
+  }
+
+  let failedToReadError = false;
+  for (const validator of confirmedValidators) {
+    let storageLocations;
+    try {
+      storageLocations = await validatorAnnounce.getAnnouncedStorageLocations([
+        validator,
+      ]);
+      const s3StorageLocation = storageLocations[0][0];
+
+      const s3Validator = await S3Validator.fromStorageLocation(
+        s3StorageLocation,
+      );
+
+      const latestCheckpointIndex =
+        await s3Validator.getLatestCheckpointIndex();
+
+      logBlue(
+        `✅ Validator ${validator} announced\nstorage location: ${s3StorageLocation}\nlatest checkpoint index: ${latestCheckpointIndex}`,
```

Comment:
an additional thing to print that would be nice:
our storage location format in the announcement is a bespoke uri format, but often when debugging I navigate to the s3 bucket's webpage URL to poke around. It looks like this: https://dsrv-hyperlane-v3-validator-signatures-validator1-ethereum.s3.eu-central-1.amazonaws.com

Could you add a function to `S3Validator` to get the URL formatted like that and also log it here? sorry for the scope increase

Reply:
I'll do this as a follow up